### DISCUSSION TITLE
Simple SPARQL-star query fails

### DISCUSSION THREAD

**Discussion by pulquero** [eclipse-rdf4j/rdf4j#3949] [Q&A]
### Current Behavior

With an empty MemoryStore, the following returns no results:
```
SELECT ?t { bind(<< <http://whatever/a> <http://whatever/val> 1 >> as ?t) }
```
I would have thought this should work in the same way I can bind a constant IRI or literal.

### Expected Behavior

t should be returned with the constant triple value  << <http://whatever/a> <http://whatever/val> 1 >>.

### Steps To Reproduce

```
        Repository repo = new SailRepository(new MemoryStore());
        repo.init();
        RepositoryConnection con = repo.getConnection();
    	String q ="SELECT ?t { bind(<< <http://whatever/a> <http://whatever/val> 1 >> as ?t) }";
        TupleQueryResult res = con.prepareTupleQuery(QueryLanguage.SPARQL, q).evaluate();
        assertTrue(res.hasNext());
```

### Version

3.7.6

### Anything else?

Query tree is
```
QueryRoot
   Projection
      ProjectionElemList
         ProjectionElem "t"
      Extension
         TripleRef
            Var (name=_const_1e762b06_uri, value=http://whatever/a, anonymous)
            Var (name=_const_5997dc86_uri, value=http://whatever/val, anonymous)
            Var (name=_const_31_lit_5fc8fb17_0, value="1"^^<http://www.w3.org/2001/XMLSchema#integer>, anonymous)
            Var (name=_anon_93abe671_2aa3_48d8_a0e6_6f2fc4a51c90, anonymous)
         ExtensionElem (t)
            Var (name=_anon_93abe671_2aa3_48d8_a0e6_6f2fc4a51c90, anonymous)
```
I feel this should really be a ValueExprTripleRef instead of a TripleRef.

This query does work however:
```
"SELECT (<< <http://whatever/a> <http://whatever/val> 1 >> as ?t) {}
```
query tree
```
QueryRoot
   Projection
      ProjectionElemList
         ProjectionElem "t"
      Extension
         SingletonSet
         ExtensionElem (t)
            ValueExprTripleRef
               Var (name=_const_1e762b06_uri, value=http://whatever/a, anonymous)
               Var (name=_const_5997dc86_uri, value=http://whatever/val, anonymous)
               Var (name=_const_31_lit_5fc8fb17_0, value="1"^^<http://www.w3.org/2001/XMLSchema#integer>, anonymous)
```

**Comment by damyan-ognyanov** [comment]
The behaviour is correct with respect to the proposed semantics of SPARQL-star. e.g. see https://arxiv.org/pdf/1406.3399.pdf section 5.2.5

hth,
Damyan

**Reply by pulquero** [reply]
I don't quite see that, you might need to spell it out for me.

Wouldn't you agree it is very counterintuitive given the VALUES clause does work as I would expect:
```
SELECT ?t
{
    values ?t {<< <http://whatever/a> <http://whatever/val> 1 >>} 
} 
```
Also,
```
SELECT ?t ?x
{
    values ?t {<< <http://whatever/a> <http://whatever/val> 1 >>} 
    bind(?t as ?x)
}
```

So it seems like the worst of all possible worlds to me:
1) constant triples don't behave like other constants,
2) VALUES ?x {x} doesn't perform consistently like BIND(x as ?x),
3) BIND of a triple variable that has a constant value is not the same as BIND of that constant value.

Seems like it breaks the internal consistency of the algebra.

**Reply by damyan-ognyanov** [reply]
The section 5.3 in the reffered document describes the evaluation semantics. The case covering the semantics of `BIND` refer further to definition 13 from section 4 (case 2) and as I interpred it - the solutions returned/matched by the expression, must be part(e.g. present) in the dataset (the said graph G, reffered in the section). 
So evaluating `bind` with fixed components of a triple pattern should only return( e.g. bind) values that are present in that dataset, if these are not present (e.g. not used as subject or object of an rdf-star triple expression), no value should be bound from BIND.
So, when evaluating the above expression(query) over an empy dataset, it should not return any solutions because there are no such nodes that could be matched.

**Reply by afs** [reply]
https://arxiv.org/pdf/1406.3399.pdf (RDF*) != [Final community group report](https://www.w3.org/2021/12/rdf-star.html#sparql-star) (RDF-star).

1406.3399 has been updated - see the red text at the top of the online version.

**Reply by hmottestad** [reply]
### Old - [1406.3399.pdf](https://arxiv.org/pdf/1406.3399.pdf)
<img width="1004" alt="Screenshot 2022-06-01 at 12 35 29" src="https://user-images.githubusercontent.com/797185/171385517-40784d0a-611a-4927-928c-8e30630467b7.png">

<img width="1039" alt="Screenshot 2022-06-01 at 12 33 15" src="https://user-images.githubusercontent.com/797185/171385246-8fe7cefb-f0d2-496c-99d0-aaa012b4a8ed.png">

### New - [Final community group report](https://www.w3.org/2021/12/rdf-star.html#semantics)
<img width="1083" alt="Screenshot 2022-06-01 at 12 33 59" src="https://user-images.githubusercontent.com/797185/171385270-66fc7bbb-b408-448f-9a25-e81d231a719e.png">

**Reply by hmottestad** [reply]
Somewhat related: https://github.com/eclipse/rdf4j/issues/2920

**Reply by hmottestad** [reply]
@jeenbroekstra do you know where we stand on our implementation of the final community group report specification for SPARQL-star?

**Reply by damyan-ognyanov** [reply]
Indeed, it may be obsolete, but that is the document(or a version of it, from around the begining of 2020) on which the current implementation is based on, including the semantincs of `BIND`, discused in this thread.

**Reply by pulquero** [reply]
OK, I get it now (I think), thanks. So, in essence, the issue revolves around how to interpret BIND(<<?s ?p ?o>> as ?t), i.e. triple with some non constant components. The old version says you bind values by searching in the graph, and I think this aligns with the PG evaluation mode where a triple* pattern infers the existence of any embedded triples, and is concerned with being compatible with reification where the embedded triple would of course exist as a triple in the graph. I have an SA mindset where BIND(<<?s ?p ?o>> as ?t) is in error if any of ?s, ?p, ?o are unbound - <<?s ?p ?o>> acts like a function. In the new spec, it seems it tries to avoid all these issues by simply having BIND() not support triples and having an explicit TRIPLE() function.

Is this about right?

**Reply by pulquero** [reply]
Further, I assume "Extensible Value Testing" means BIND(<< <http://whatever/a> <http://whatever/val> 1 >> as ?t) in the new spec could still behave like the old spec.

**Reply by afs** [reply]
That's the idea - it is SA, there is a new RDF term for a quoted triple. The rest follows.

> by simply having BIND() not support triples and having an explicit TRIPLE() function.

as in "not support triple matching".

The difference between `<<... >>` as an expression for a quoted triple, and the function `TRIPLE` is that `TRIPLE` allows any expression whereas `<<>>` only permits an RDF term or a variable in each position. e.g `TRIPLE(IRI(?str), :p, ?x+1)`.

https://www.w3.org/2021/12/rdf-star.html#quoted-triple-expression

**Reply by abrokenjester** [reply]
> @jeenbroekstra do you know where we stand on our implementation of the final community group report specification for SPARQL-star?

Like @damyan-ognyanov mentioned, the current implementation is based on an earlier version, and I believe we have not really done much futher work on this since then - with myself changing jobs and the Ontotext team (who contributed most of the original implementation) shifting focus to other things.
